File: Cpp-Src/Main/network.cpp
#include "egghead_comp.hpp"

void Network::forward(vector<double> inputs)
{
    for (int index = 0; index < inputs.size(); index++)
        *this->inputLayer->at(index)->stateValue = inputs[index];
    for (int index = 0; index < this->inputLayer->size(); index++)
        this->inputLayer->at(index)->pipeLine->forwardResult = this->inputLayer->at(index)->stateValue;
    for (int index = 0; index < this->layers->size(); index++)
        this->layers->at(index)->forward();
    for (int index = 0; index < this->outputLayer->size(); index++)
        this->outputLayer->at(index)->stateValue = this->outputLayer->at(index)->pipeLine->forwardResult;
}

void Network::backward(vector<double> expected)
{
    double constant = 1.0 / expected.size();
    for (int index = 0; index < expected.size(); index++)
    {
        double paramValue = (-1.0 * constant * (expected[index] - *this->outputLayer->at(index)->stateValue));
        *this->outputLayer->at(index)->pipeLine->backwardResult = paramValue;
    }
    
    int indexLayer = this->layers->size() - 1;
    while (indexLayer >= 0)
    {
        this->layers->at(indexLayer--)->backward();
    }
}

Network::Network(double newLearningRate)
{
    learningRate = new double(newLearningRate);
    layers = new vector<Layer *>();
    inputLayer = nullptr;
    outputLayer = nullptr;
}

Network::~Network()
{
    delete learningRate;
    delete layers;
    delete inputLayer;
    delete outputLayer;
}

void Network::buildNetwork()
{
    cout << "Please implement BuildNetwork in the child class." << endl;
}

void Network::trainNetwork(int iterations, vector<vector<double>*>* cases)
{
    cout << "Please implement TrainNetwork in the child class." << endl;
}

void Network::connectPipeline(Node * sourceNode, Node * targetNode)
{
    Pipe * pipeline = new Pipe();
    sourceNode->frontPipes->push_back(pipeline);
    targetNode->backPipes->push_back(pipeline);
}

void Network::connectStateToNode(State * sourceState, Node * targetNode)
{
    if (sourceState->pipeLine == nullptr)
    {
        Pipe * pipeline = new Pipe();
        sourceState->pipeLine = pipeline;
        targetNode->backPipes->push_back(pipeline);
    }
    else
    {
        targetNode->backPipes->push_back(sourceState->pipeLine);
    }
}

void Network::connectNodeToState(Node* sourceNode, State * targetState)
{
    if (targetState->pipeLine == nullptr)
    {
        Pipe* pipeline = new Pipe();
        sourceNode->frontPipes->push_back(pipeline);
        targetState->pipeLine = pipeline;
    }
    else
    {
        sourceNode->frontPipes->push_back(targetState->pipeLine);
    }
}

vector<double> Network::runNormal(vector<double> inputs, bool returnRaw)
{
    this->forward(inputs);
    vector<double> ans;

    for (int index = 0; index < this->outputLayer->size(); index++)
    {
        if (returnRaw)
        {
            ans.push_back(*this->outputLayer->at(index)->stateValue);
        }
        else
        {
            ans.push_back(0.5 <= *this->outputLayer->at(index)->stateValue ? 1.0 : 0.0);
        }
    }

    return ans;
}

void Network::runForward(vector<double> inputs)
{
    this->forward(inputs);
}

void Network::runBackward(vector<double> expected)
{
    this->backward(expected);
}

File: Csharp-Src/Csharp-Src/Network.cs
using System;
using System.Collections.Generic;

namespace Csharp_Src
{
    public class Network
    {
        public double LearningRate { get; }
        public List<Layer> Layers { get; set; }
        public List<State> InputLayer { get; set; }
        public List<State> OutputLayer { get; set; }

        public Network(double learningRate = 0.01, List<Layer> layers = null)
        {
            this.LearningRate = learningRate;
            this.Layers = layers != null ? layers : new List<Layer>();
            this.InputLayer = null;
            this.OutputLayer = null;
        }

        public virtual void BuildNetwork()
        {
            throw new Exception("Please implement BuildNetwork in the child class.");
        }

        public virtual void TrainNetwork(int iterations, List<List<double>> cases)
        {
            throw new Exception("Please implement TrainNetwork in the child class.");
        }

        public void ConnectPipeline(Node sourceNode, Node targetNode)
        {
            Pipe pipeline = new Pipe();
            sourceNode.FrontPipes.Add(pipeline);
            targetNode.BackPipes.Add(pipeline);
        }

        public void ConnectStateToNode(State sourceState, Node targetNode)
        {
            if (sourceState.Pipeline == null)
            {
                Pipe pipeline = new Pipe();
                sourceState.Pipeline = pipeline;
                targetNode.BackPipes.Add(pipeline);
            }
            else
            {
                targetNode.BackPipes.Add(sourceState.Pipeline);
            }
        }

        public void ConnectNodeToState(Node sourceNode, State targetState)
        {
            if (targetState.Pipeline == null)
            {
                Pipe pipeline = new Pipe();
                sourceNode.FrontPipes.Add(pipeline);
                targetState.Pipeline = pipeline;
            }
            else
            {
                sourceNode.FrontPipes.Add(targetState.Pipeline);
            }
        }

        public List<double> RunNormal(List<double> inputs, bool returnRaw = false)
        {
            this.Forward(inputs);
            List<double> ans = new List<double>();

            foreach (var outputState in this.OutputLayer)
            {
                if (returnRaw)
                    ans.Add(outputState.StateValue);
                else
                    ans.Add(0.5 <= outputState.StateValue ? 1 : 0);
            }

            return ans;
        }

        public void RunForward(List<double> inputs)
        {
            this.Forward(inputs);
        }

        private void Forward(List<double> inputs)
        {
            if (inputs.Count != this.InputLayer.Count)
                throw new Exception("Inputs don\'t aline with input states. Please check again.");

            for (int index = 0; index < inputs.Count; index++)
                this.InputLayer[index].StateValue = inputs[index];
            foreach (var inputState in this.InputLayer)
                inputState.Pipeline.ForwardResult = inputState.StateValue;
            foreach (var layer in this.Layers)
                layer.Forward();
            foreach (var outputState in this.OutputLayer)
                outputState.StateValue = outputState.Pipeline.ForwardResult;
        }

        public void RunBackward(List<double> expected)
        {
            this.Backward(expected);
        }

        private void Backward(List<double> expected)
        {
            if (expected.Count != this.OutputLayer.Count)
                throw new Exception("Expected and actual length isn\'t valid.");
            
            double constant = 1.0 / expected.Count;
            
            for (int index = 0; index < expected.Count; index++)
                this.OutputLayer[index].Pipeline.BackwardResult = (-1.0 * constant * (expected[index] - this.OutputLayer[index].StateValue));
            
            int indexLayer = this.Layers.Count - 1;

            while (indexLayer >= 0)
                this.Layers[indexLayer--].Backward();
        }
    }
}
